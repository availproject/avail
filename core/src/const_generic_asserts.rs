use core::mem::size_of;

pub struct UsizeNonZero<const N: usize>;
impl<const N: usize> UsizeNonZero<N> {
	#[allow(dead_code)]
	pub const OK: () = assert!(N != 0, "must be non-zero");
}

pub struct UsizeEven<const N: usize>;
impl<const N: usize> UsizeEven<N> {
	#[allow(dead_code)]
	pub const OK: () = assert!(N % 2 == 0, "must be even");
}

pub struct USizeSafeCastToU32<const N: usize>;
impl<const N: usize> USizeSafeCastToU32<N> {
	#[allow(dead_code)]
	pub const OK: () = assert!(
		size_of::<usize>() <= size_of::<u32>() || N <= u32::MAX as usize,
		"must be safe to cast to u32"
	);
}

pub struct USizeGreaterOrEq<const N: usize, const M: usize>;
impl<const N: usize, const M: usize> USizeGreaterOrEq<N, M> {
	#[allow(dead_code)]
	pub const OK: () = assert!(N >= M, "must be greater or equal");
}

#[cfg(test)]
mod tests {
	#[test]
	fn non_zero() {
		let t = trybuild::TestCases::new();
		t.pass("tests/const_generics/usize_non_zero.rs");
		t.compile_fail("tests/const_generics/usize_non_zero_fail.rs");
	}

	#[test]
	fn even() {
		let t = trybuild::TestCases::new();
		t.pass("tests/const_generics/usize_even.rs");
		t.compile_fail("tests/const_generics/usize_even_fail.rs");
	}

	#[test]
	fn safe_cast_to_u32() {
		let t = trybuild::TestCases::new();
		t.pass("tests/const_generics/usize_safe_cast_to_u32.rs");
		t.compile_fail("tests/const_generics/usize_safe_cast_to_u32_fail.rs");
	}

	#[test]
	fn greater_or_eq() {
		let t = trybuild::TestCases::new();
		t.pass("tests/const_generics/usize_greater_or_eq.rs");
		t.compile_fail("tests/const_generics/usize_greater_or_eq_fail.rs");
	}
}
