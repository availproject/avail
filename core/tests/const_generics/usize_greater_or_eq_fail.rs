use avail_core::const_generic_asserts::USizeGreaterOrEq;

fn const_generic_ge<const N: usize, const M: usize> () {
	let () = USizeGreaterOrEq::<N,M>::OK;
}

fn main () {
	const_generic_ge::<1, 2>();
}
