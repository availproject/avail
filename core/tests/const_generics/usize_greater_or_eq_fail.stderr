error[E0080]: evaluation of `avail_core::const_generic_asserts::USizeGreaterOrEq::<1, 2>::OK` failed
 --> src/const_generic_asserts.rs
  |
  |     pub const OK: () = assert!(N >= M, "must be greater or equal");
  |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'must be greater or equal', $DIR/src/const_generic_asserts.rs:27:24
  |
  = note: this error originates in the macro `$crate::panic::panic_2021` which comes from the expansion of the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn const_generic_ge::<1, 2>`
 --> tests/const_generics/usize_greater_or_eq_fail.rs:8:2
  |
8 |     const_generic_ge::<1, 2>();
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
